Fix TCPStoreTest and improve tcputils::connect()

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30354 Fix TCPStoreTest and improve tcputils::connect()**

TCPStoreTest would timeout since the TCPStore constructor for the
server would block the main thread waiting for workers. The workers themselves
were spawned later on once the server store is created. As a result, this test
would always timeout.

To fix the test, I moved the server store to a thread so that the workers can
register with the server in parallel.

In addition to this made a few improvements to tcputils::connect. When
tcputils::connect() encountered an exception, it always looked at `errno` for
the error code. In some cases `errno` could be overwritten and the real error
code would be stored in `std::system_error`. As a result, I've modified the
code to look at the error code in `std::system_error` if we catch an exception
of that type.

Differential Revision: [D18668454](https://our.internmc.facebook.com/intern/diff/D18668454/)

Commit:
Fix TCPStoreTest and improve tcputils::connect()

TCPStoreTest would timeout since the TCPStore constructor for the
server would block the main thread waiting for workers. The workers themselves
were spawned later on once the server store is created. As a result, this test
would always timeout.

To fix the test, I moved the server store to a thread so that the workers can
register with the server in parallel.

In addition to this made a few improvements to tcputils::connect. When
tcputils::connect() encountered an exception, it always looked at `errno` for
the error code. In some cases `errno` could be overwritten and the real error
code would be stored in `std::system_error`. As a result, I've modified the
code to look at the error code in `std::system_error` if we catch an exception
of that type.

Differential Revision: [D18668454](https://our.internmc.facebook.com/intern/diff/D18668454/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -1,26 +1,36 @@
-  c10d::PrefixStore serverStore(prefix, serverTCPStore);
+  auto serverThread =
+      std::thread([&serverTCPStore, &serverStore, &prefix, &numWorkers, &port] {
+        serverTCPStore = std::make_unique<c10d::TCPStore>(
```
Should we also pass in a timeout to the constructor? That way if TCPStore ever blocks in `waitForWorkers` we can terminate quickly with an error. A good default could be ~10 seconds

pritamdamania87 replied:
Went with 30s to be conservative.

Commit:
Update on "Fix TCPStoreTest and improve tcputils::connect()"


TCPStoreTest would timeout since the TCPStore constructor for the
server would block the main thread waiting for workers. The workers themselves
were spawned later on once the server store is created. As a result, this test
would always timeout.

To fix the test, I moved the server store to a thread so that the workers can
register with the server in parallel.

In addition to this made a few improvements to tcputils::connect. When
tcputils::connect() encountered an exception, it always looked at `errno` for
the error code. In some cases `errno` could be overwritten and the real error
code would be stored in `std::system_error`. As a result, I've modified the
code to look at the error code in `std::system_error` if we catch an exception
of that type.

Differential Revision: [D18668454](https://our.internmc.facebook.com/intern/diff/D18668454/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -131,6 +131,71 @@ std::pair<int, PortType> listen(PortType port) {
+    *anyRefused = true;
+  }
+  // ECONNRESET happens if the server's listen backlog is exhausted.
+  if (error_code == ECONNRESET) {
```
can `error_code` ever be something else? Should we throw if we get an error_code that is not one of these two?

rohan-varma reviewed (approved):
Looks good, approve to unblock although @pietern should likely take a look at the error handling as well.

Also, I don't think we can use `make_unique` in CI:

`Nov 25 22:25:42 /var/lib/jenkins/workspace/torch/lib/c10d/test/TCPStoreTest.cpp:21:26: error: no template named 'make_unique' in namespace 'std'; did you mean 'c10::guts::make_unique'?
Nov 25 22:25:42         serverTCPStore = std::make_unique<c10d::TCPStore>(`

pietern commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -1,26 +1,36 @@
 #include <c10d/test/StoreTestCommon.hpp>
 
+#include <gtest/gtest.h>
```
Move after stdlib includes.

pietern commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -1,26 +1,36 @@
+        c10d::test::set(*serverStore, "key2", "value2");
+        c10d::test::check(*serverStore, "key0", "value0");
+        c10d::test::check(*serverStore, "key1", "value1");
+        c10d::test::check(*serverStore, "key2", "value2");
```
This thread terminates once this has run. I suppose you're relying on the instance and by extension the server thread to be kept alive by the `unique_ptr` that holds it?

I think that, alternatively, we should consider decouple starting the TCP server from the waiting for all workers. This logic is too complicated for what we should test here.

pietern replied:
> I'm not sure I followed this completely. The server TCP store is kept alive by the unique_ptr on line 16.

Understood.

> IIUC, is your suggestion to change the TCPStore constructor to not wait for workers and have a separate method for this?

Yes, I think this would a) simplify this test (and similar patterns), because you no longer need a thread to launch a thread, and b) it would allow for letting the server pick a free port automatically (now you have to wait for all workers to connect, so it can't be ambiguous).

pietern reviewed (commented):
The templated error handlers in combination with the exceptions for control flow make me uneasy.

There is an opportunity to simplify this a ton by removing exceptions altogether IMO.

Did this test never work or get run in the first place?

rohan-varma reviewed (changes requested):
Remove stamp until @pietern 's comments are addressed.

Commit:
Update on "Fix TCPStoreTest and improve tcputils::connect()"


TCPStoreTest would timeout since the TCPStore constructor for the
server would block the main thread waiting for workers. The workers themselves
were spawned later on once the server store is created. As a result, this test
would always timeout.

To fix the test, I moved the server store to a thread so that the workers can
register with the server in parallel.

In addition to this made a few improvements to tcputils::connect. When
tcputils::connect() encountered an exception, it always looked at `errno` for
the error code. In some cases `errno` could be overwritten and the real error
code would be stored in `std::system_error`. As a result, I've modified the
code to look at the error code in `std::system_error` if we catch an exception
of that type.

Differential Revision: [D18668454](https://our.internmc.facebook.com/intern/diff/D18668454/)

[ghstack-poisoned]

pritamdamania87 commented:
> The templated error handlers in combination with the exceptions for control flow make me uneasy.

Fixed the template

> There is an opportunity to simplify this a ton by removing exceptions altogether IMO.

This might be tricky to do since the `SYSCHECK_ERR_RETURN_NEG1` macros throw exceptions as well.

> Did this test never work or get run in the first place?

I tried to look at the history for when this test was added and it seems like this test should have never worked.

pietern reviewed (approved):
Thanks, @pritamdamania87. Let's merge to unblock and revisit the exceptions for control flow thing and starting a store without waiting for all workers.

Commit:
Update on "Fix TCPStoreTest and improve tcputils::connect()"


TCPStoreTest would timeout since the TCPStore constructor for the
server would block the main thread waiting for workers. The workers themselves
were spawned later on once the server store is created. As a result, this test
would always timeout.

To fix the test, I moved the server store to a thread so that the workers can
register with the server in parallel.

In addition to this made a few improvements to tcputils::connect. When
tcputils::connect() encountered an exception, it always looked at `errno` for
the error code. In some cases `errno` could be overwritten and the real error
code would be stored in `std::system_error`. As a result, I've modified the
code to look at the error code in `std::system_error` if we catch an exception
of that type.

Differential Revision: [D18668454](https://our.internmc.facebook.com/intern/diff/D18668454/)

[ghstack-poisoned]
